### PR TITLE
Better help message for existing applications

### PR DESCRIPTION
**Pull request by SimonRichardson**
When deploying an existing application again via `juju deploy` the help
message is a bit sad, as it doesn't explain how you get out of that
position and move forward.

The following aim so to do that by using an error type all the way from
state and catching that error code in the charm deployer.

Interestingly the state tests don't change as we're backwards compatible
with the old error message (nice!) and the charm one isn't exactly easy
to trigger.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy mysql
$ juju deploy mysql
Located charm "mysql" in charm-hub, revision 58
Deploying "mysql" from charm-hub charm "mysql", revision 58 in channel stable
ERROR cannot add application "mysql": application already exists:
deploy application using an alias name, or use remove-application to remove the existing one and try again
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1915223

### CODE REVIEW

File: cmd/juju/application/deployer/charm.go [hunk 1]
```diff
@@ -252,7 +253,20 @@ func (d *deployCharm) deploy(
...
 		EndpointBindings: d.bindings,
 	}
-	return errors.Trace(deployAPI.Deploy(args))
+
+	err = deployAPI.Deploy(args)
+	if err == nil {
+		return nil
+	}
+
+	if pErr, ok := errors.Cause(err).(*params.Error); ok {
```
Comment (achilleasa) [inline cmd/juju/application/deployer/charm.go:262]: How about moving this check into the client and have that return an `errors.AlreadyExists` that can be tested for and wrapped here? I would prefer we didn't deal with `params.Error` at this level.

Comment (SimonRichardson) [reply cmd/juju/application/deployer/charm.go:262]: Sure makes sense

File: cmd/juju/application/deployer/charm.go [hunk 2]
```diff
@@ -259,11 +258,8 @@ func (d *deployCharm) deploy(
 		return nil
 	}
 
-	if pErr, ok := errors.Cause(err).(*params.Error); ok {
-		switch pErr.Code {
-		case params.CodeAlreadyExists:
-			return errors.Wrapf(err, errors.Errorf("\ndeploy application using an alias name, or use remove-application to remove the existing one and try again"), err.Error())
-		}
+	if errors.IsAlreadyExists(err) {
+		return errors.Wrapf(err, errors.Errorf("\ndeploy application using an alias name, or use remove-application to remove the existing one and try again"), err.Error())
```
Comment (achilleasa) [inline cmd/juju/application/deployer/charm.go:262]: Minor nit; I think you can get rid of the `\n` so it reads as one line after the `:` when the error is printed out.

**Review by achilleasa — APPROVED** [review]
LGTM; just a minor nit for the line-feed in the error (feel free to ignore)

File: cmd/juju/application/deployer/charm.go (hunk 2 above)
Comment (SimonRichardson) [reply cmd/juju/application/deployer/charm.go:262]: The error message is huge otherwise...

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**